modules/service/src/main/scala/lucuma/odb/service/ItcService.scala

```diff
@@ -0,0 +1,360 @@
+import lucuma.odb.service.Services.Syntax.*
+import lucuma.odb.util.Codecs.*
+import skunk.*
+import skunk.codec.text.text
+import skunk.implicits.*
+
+import java.security.MessageDigest
+import java.util.HexFormat
+
+sealed trait ItcService[F[_]] {
+
+  def lookup(
+    programId:     Program.Id,
+    observationId: Observation.Id
+  )(using NoTransaction[F]): F[Either[ItcService.Error, ItcService.Success]]
+
+}
+
+object ItcService {
+
+  sealed trait Error {
+    def format: String
+  }
+
+  object Error {
+
+    case class ObservationNotFound(
+      programId:     Program.Id,
+      observationId: Observation.Id
+    ) extends Error {
+      def format: String =
+        s"Observation '$observationId' in program '$programId' not found."
+    }
+
+    case class MissingParams(
+      missing: NonEmptyList[GeneratorParamsService.MissingData]
+    ) extends Error {
+      def format: String = {
+        val params = missing.map { m =>
+          s"${m.targetId.fold("") { tid => s"(target $tid) " }}${m.paramName}"
+        }.intercalate(", ")
+        s"ITC cannot be queried until the following parameters are defined: $params"
+      }
+    }
+
+    case class RemoteServiceErrors(
+      problems: NonEmptyList[(Target.Id, String)]
+    ) extends Error {
+      def format: String = {
+        val ps = problems.map { case (tid, msg) => s"Target '$tid': $msg"}
+        s"ITC returned errors: ${ps.intercalate(", ")}"
+      }
+    }
+  }
+
+  case class TargetResult(
+    targetId: Target.Id,
+    input:    SpectroscopyIntegrationTimeInput,
+    value:    IntegrationTime,
+    versions: ItcVersions
+  ) {
+    def totalTime: Option[TimeSpan] = {
+      val total = BigInt(value.exposureTime.toMicroseconds) * value.exposures.value
+      Option.when(total.isValidLong)(TimeSpan.fromMicroseconds(total.longValue)).flatten
+    }
+  }
+
+  object TargetResult {
+
+    given Order[TargetResult] =
+      Order.by { s => (s.totalTime, s.targetId) }
+
+    import lucuma.odb.json.time.query.given
+
+    given Encoder[TargetResult] =
+      Encoder.instance { s =>
+        Json.obj(
+          "targetId"      -> s.targetId.asJson,
+          "exposureTime"  -> s.value.exposureTime.asJson,
+          "exposures"     -> s.value.exposures.value.asJson,
+          "signalToNoise" -> s.value.signalToNoise.asJson
+        )
+      }
+  }
+
+  case class AsterismResult(
+    value: Zipper[TargetResult]
+  )
+
+  object AsterismResult {
+
+    given Encoder[AsterismResult] =
+      Encoder.instance { rs =>
+        Json.obj(
+          "result" -> rs.value.focus.asJson,
+          "all"    -> rs.value.toList.asJson
+        )
+      }
+
+  }
+
+  case class Success(
+    params: GeneratorParams,
+    result: AsterismResult
+  )
+
+  private val hex = HexFormat.of
+
+  private def hash(input: SpectroscopyIntegrationTimeInput): String =
+    hex.formatHex(
+      MessageDigest
+        .getInstance("MD5")
+        .digest(input.asJson.noSpaces.getBytes("UTF-8"))
+    )
```

Comment:
Here is the hash.  Any better ideas? This is simple but it would be better not to first encode everything to JSON.  On the other hand, if the ITC server input changes the JSON changes so there's no way to overlook it when computing the hash this way.

Reply:
This is the safest way to do it, so I think it's fine to start here. May never be an issue.